hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java

```diff
@@ -76,34 +84,179 @@
+      ConfigurationSource conf, int threadPoolSize, int queueLimit) {
     this.containerSet = cset;
     this.conf = conf;
+    this.executor = new ThreadPoolExecutor(
+        0, threadPoolSize, 60, TimeUnit.SECONDS,
+        new LinkedBlockingQueue<>(),
+        new ThreadFactoryBuilder().setDaemon(true)
+            .setNameFormat("DeleteBlocksCommandHandlerThread-%d")
+            .build());
+    this.deleteCommandQueues = new LinkedBlockingQueue<>(queueLimit);
+    handlerThread = new Daemon(new DeleteCmdWorker());
+    handlerThread.start();
   }
 
   @Override
   public void handle(SCMCommand command, OzoneContainer container,
       StateContext context, SCMConnectionManager connectionManager) {
-    cmdExecuted = false;
-    long startTime = Time.monotonicNow();
-    ContainerBlocksDeletionACKProto blockDeletionACK = null;
+    if (command.getType() != SCMCommandProto.Type.deleteBlocksCommand) {
+      LOG.warn("Skipping handling command, expected command "
+              + "type {} but found {}",
+          SCMCommandProto.Type.deleteBlocksCommand, command.getType());
+      return;
+    }
+
     try {
-      if (command.getType() != SCMCommandProto.Type.deleteBlocksCommand) {
-        LOG.warn("Skipping handling command, expected command "
-                + "type {} but found {}",
-            SCMCommandProto.Type.deleteBlocksCommand, command.getType());
-        return;
+      DeleteCmdInfo cmd = new DeleteCmdInfo((DeleteBlocksCommand) command,
+          container, context, connectionManager);
+      deleteCommandQueues.add(cmd);
+    } catch (IllegalStateException e) {
+      LOG.warn("Command is discarded because of the command queue is full");
+      return;
+    }
+  }
+
+  /**
+   * A delete command info.
+   */
+  public static final class DeleteCmdInfo {
+    private DeleteBlocksCommand cmd;
+    private StateContext context;
+    private OzoneContainer container;
+    private SCMConnectionManager connectionManager;
+
+    public DeleteCmdInfo(DeleteBlocksCommand command, OzoneContainer container,
+        StateContext context, SCMConnectionManager connectionManager) {
+      this.cmd = command;
+      this.context = context;
+      this.container = container;
+      this.connectionManager = connectionManager;
+    }
+    public DeleteBlocksCommand getCmd() {
+      return this.cmd;
+    }
+    public StateContext getContext() {
+      return this.context;
+    }
+    public OzoneContainer getContainer() {
+      return this.container;
+    }
+    public SCMConnectionManager getConnectionManager() {
+      return this.connectionManager;
+    }
+  }
+
+  /**
+   * Process delete commands.
+   */
+  public final class DeleteCmdWorker implements Runnable {
+
+    @Override
+    public void run() {
+      while (true) {
+        while (!deleteCommandQueues.isEmpty()) {
+          DeleteCmdInfo cmd = deleteCommandQueues.poll();
+          try {
+            processCmd(cmd.getCmd(), cmd.getContainer(), cmd.getContext(),
```

Comment:
thanks for the work ! NIT, can we use `DeleteCmdInfo` as the only one parameter of `processCmd`

Reply:
Sure.